pkg/frontend/querymiddleware/results_cache.go

```diff
@@ -232,30 +235,30 @@ func isAtModifierCachable(r Request, maxCacheTime int64, logger log.Logger) bool
 	expr = promql.PreprocessExpr(expr, timestamp.Time(r.GetStart()), timestamp.Time(r.GetEnd()))
 
 	end := r.GetEnd()
-	atModCachable := true
+	cachable := true
+	check := func(ts *int64, offset time.Duration) error {
+		if offset < 0 {
+			cachable = false
+			return errNegativeOffset
+		}
+		if ts != nil && (*ts > end || *ts > maxCacheTime) {
+			cachable = false
+			return errAtModifierAfterEnd
+		}
+		return nil
+	}
+
 	parser.Inspect(expr, func(n parser.Node, _ []parser.Node) error {
 		switch e := n.(type) {
 		case *parser.VectorSelector:
-			if e.Timestamp != nil && (*e.Timestamp > end || *e.Timestamp > maxCacheTime) {
-				atModCachable = false
-				return errAtModifierAfterEnd
-			}
-		case *parser.MatrixSelector:
```

Comment:
Can you confirm the reason why you removed it is bacause `VectorSelector` is its child, so will be recursive hit anyway?

Reply:
Yes.